File: roblox_plugin/src/Server/formatTypes/mask.ts
//!native
import type * as ProcessFormats from "../processFormats";
import SendMessagesThread from "../sendMessages";
import ReceiveMessagesThread from "../receiveMessages";

interface MaskFormatData {
	type: "mask";
	mask: string;
	invert?: boolean;
	data: ProcessFormats.FormatData;
}
export function read(
	receiveThread: ReceiveMessagesThread,
	args: defined[],
	formatData: MaskFormatData,
	masks: Map<string, boolean>,
) {
	const mask = formatData.mask;
	let condition = !!masks.get(mask);
	if ("invert" in formatData && formatData.invert) {
		condition = !condition;
	}
	if (condition) {
		ReceiveMessagesThread.parse(receiveThread, args, formatData.data, masks);
	}
}
export function write(
	sendThread: SendMessagesThread,
	args: defined[],
	argsCount: number,
	formatData: MaskFormatData,
	masks: Map<string, boolean>,
): number {
	const mask = formatData.mask;
	let condition = !!masks.get(mask);
	if ("invert" in formatData && formatData.invert) {
		condition = !condition;
	}
	if (condition) {
		argsCount = SendMessagesThread.parse(sendThread, args, argsCount, formatData.data, masks);
	}
	return argsCount;
}

File: blender_roblox_sync/roblox_interface/server/format_types/mask.py
def read(self, format_data, **masks):
	mask = format_data["mask"]
	condition = (mask in masks) and masks[mask]
	if ("invert" in format_data) and format_data["invert"]:
		condition = not condition
	if condition:
		return self.parse(format_data["data"], **masks)
	else:
		return ()

def write(self, args, args_count, format_data, **masks):
	mask = format_data["mask"]
	condition = (mask in masks) and masks[mask]
	if ("invert" in format_data) and format_data["invert"]:
		condition = not condition
	if condition:
		args_count = self.parse(args, args_count, format_data["data"], **masks)
	return args_count